### PR TITLE
Isolate digest partition scheduling/maintenance runs.

### LINKED ISSUE #3227: Better handle partial failures in digest scheduler loop.
Failure to schedule an individual partition (e.g. due to failure to acquire the scheduler lock for that partition) can cause the loop to exit prematurely and prevent other partitions that would not be affected by the failing partitions to be scheduled.

### PR DESCRIPTION
**Pull request by tkaemming**
This ensures that failing to perform scheduling or maintenance on a partition does not cause the remainder of the unprocessed partitions to fail.

Fixes GH-3227.

@getsentry/infrastructure

### CODE REVIEW

File: src/sentry/digests/backends/redis.py [hunk 1]
```diff
@@ -188,185 +188,205 @@ def add(self, key, record, increment_delay=None, maximum_delay=None):
...
 
                 # If we retrieved less than the chunk size of items, we don't
                 # need try to retrieve more items.
-                if len(items) < fetch_size:
+                if len(items) < chunk:
                     break
             else:
                 raise RuntimeError('loop exceeded maximum iterations (%s)' % (maximum_iterations,))
 
+    def schedule(self, deadline, chunk=1000):
```
Comment (tkaemming) [inline src/sentry/digests/backends/redis.py]: This is the substantive part of the change, the rest was just moving lines around.

File: src/sentry/digests/backends/redis.py [hunk 2]
```diff
@@ -188,185 +188,205 @@ def add(self, key, record, increment_delay=None, maximum_delay=None):
 
             return results[-2 if should_truncate else -1]
 
-    def schedule(self, deadline, chunk=1000):
-        # TODO: This doesn't lead to a fair balancing of workers, ideally each
-        # scheduling task would be executed by a different process for each
-        # host. There is also no failure isolation here, so a single shard
```
Comment (tkaemming) [inline src/sentry/digests/backends/redis.py]: > There is also no failure isolation here, so a single shard failure will cause the remainder of the shards to not be able to be scheduled.

😞

File: src/sentry/digests/backends/redis.py [hunk 3]
```diff
@@ -188,185 +188,205 @@ def add(self, key, record, increment_delay=None, maximum_delay=None):
...
+        # TODO: This doesn't lead to a fair balancing of workers, ideally each
+        # scheduling task would be executed by a different process for each
+        # host.
+        for host in self.cluster.hosts:
+            try:
+                for entry in self.__schedule_partition(host, deadline, chunk):
+                    yield entry
+            except Exception as error:
+                logger.warning(
+                    'Failed to perform scheduling for partition %r due to error: %r',
```
Comment (dcramer) [inline src/sentry/digests/backends/redis.py]: id probably keep it as an error

Comment (tkaemming) [inline src/sentry/digests/backends/redis.py]: Changed to `logger.error` in 8b166ff747c1dd7075c6755eaaf2bb1c17b25a61.

**Pull request merged**